core/src/test/java/io/methvin/watchservice/DirectoryWatcherOnDiskTest.java

```diff
@@ -426,58 +431,182 @@ public void pathsWithContexts() throws IOException, InterruptedException {
+    waitRecorderSize(3, 2);
+    
+    assertEquals(DirectoryChangeEvent.EventType.DELETE, this.recorder.events.get(0).eventType());
+    assertFalse(this.recorder.events.get(0).isDirectory());
+
+    assertEquals(DirectoryChangeEvent.EventType.DELETE, this.recorder.events.get(1).eventType());
+    assertTrue(this.recorder.events.get(1).isDirectory());
+
+    this.watcher.close();
+  }
+
+  @Test
+  public void observePreHashes() throws IOException, InterruptedException {
+    Path d1 = this.tmpDir.resolve("d1");
+    Files.createDirectory(d1);
+
+    final Path f1 = Files.createTempFile(d1, "f1-", ".dat");
+    Files.write(f1, new byte[] {counter++});
+
+    this.watcher =
+          DirectoryWatcher.builder()
+                          .paths(Arrays.asList(new Path[] {d1}))
+                          .listener(this.recorder)
+                          .fileHashing(true)
+                          .build();
+
+    this.watcher.watchAsync();
+
+    Map<Path, Hash> map = this.watcher.pathHashes();
+    assertEquals(Hash.DIRECTORY, map.get(d1));
+
+    HashCode hashCode1 = FileHasher.DEFAULT_FILE_HASHER.hash(f1);
+    assertEquals(hashCode1, map.get(f1));
+
+    try {
+      final Path f2 = Files.createTempFile(d1, "f2-", ".dat");
+      map.put(f2, null);
+      fail("Exception should be thrown");
+    } catch (UnsupportedOperationException e) {
+    }
+
+    this.watcher.close();
+  }
+
+  @Test
+  public void observeHashes() throws IOException, InterruptedException {
+    Path d1 = this.tmpDir.resolve("d1");
+    Files.createDirectory(d1);
+
+    this.watcher =
+          DirectoryWatcher.builder()
+                          .paths(Arrays.asList(new Path[] {d1}))
+                          .listener(this.recorder)
+                          .fileHashing(true)
+                          .build();
+
+    this.watcher.watchAsync();
+
+    final Path f1 = Files.createTempFile(d1, "f1-", ".dat");
+    Files.write(f1, new byte[] {counter++});
+    waitRecorderSize(3, 1);
+
+    HashCode hashCode1 = FileHasher.DEFAULT_FILE_HASHER.hash(f1);
+    assertNotSame(hashCode1, this.recorder.events.get(0).hash());
+    assertEquals(hashCode1, this.recorder.events.get(0).hash());
+
+    this.recorder.events.clear();
+    Files.write(f1, new byte[] {counter++});
+    waitRecorderSize(3, 1);
+
+    HashCode hashCode2 = FileHasher.DEFAULT_FILE_HASHER.hash(f1);
+    assertNotEquals(hashCode2, hashCode1);
+    assertNotSame(hashCode2, this.recorder.events.get(0).hash());
+    assertEquals(hashCode2, this.recorder.events.get(0).hash());
+
+    this.watcher.close();
+  }
+
+  private void waitRecorderSize(int atMost, int untilSize) {
+    await().atMost(atMost, TimeUnit.SECONDS)
+           .pollDelay(100, TimeUnit.MILLISECONDS)
+           .until(() -> this.recorder.events.size() == untilSize);
+  }
+
+  private void waitFileSize(int atMost, int length) {
+    await().atMost(atMost, TimeUnit.SECONDS).until(() -> tmpDir.toFile().listFiles().length == length);
```

Comment:
I detect that this code is problematic. According to the [Correctness (CORRECTNESS)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#correctness-correctness), [SA: Self comparison of value with itself (SA_SELF_COMPARISON)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#sa-self-comparison-of-field-with-itself-sa-field-self-comparison).
This method compares a field with itself, and may indicate a typo or a logic error. Make sure that you are comparing the right things.